lib/core/scheduler.cpp

```diff
@@ -0,0 +1,47 @@
+#include "scheduler.hpp"
+#include "wall_clock.hpp"
+
+namespace hyped::core {
+
+Scheduler::Scheduler(core::ILogger &logger, core::ITimeSource &time)
+    : logger_(logger),
+      time_(time),
+      task_queue_(),
+      task_map_()
+{
+}
+
+core::Result Scheduler::run()
+{
+  if (task_queue_.empty()) { return core::Result::kSuccess; }
+  const auto current_time = std::chrono::time_point_cast<std::chrono::microseconds>(time_.now())
+                              .time_since_epoch()
+                              .count();
+  const auto next_task_time = task_queue_.top();
+  if (current_time < next_task_time) { return core::Result::kSuccess; }
+  task_queue_.pop();
+  const auto task = task_map_.find(next_task_time);
+  if (task == task_map_.end()) {
+    logger_.log(core::LogLevel::kFatal, "Scheduled task missing callback function");
+    return core::Result::kFailure;
+  }
+  task_map_.erase(next_task_time);
```

Comment:
this is expensive.

the immediate optimisation is to do `task_map_.erase(task)` instead, as you have already found the location.

then again, i still think that this is not optimal. try and make use of the priority queue as it's designed to have elements inserted and removed.